Mage.Sets/src/mage/cards/t/TeferiTemporalPilgrim.java

```diff
@@ -0,0 +1,107 @@
+import mage.filter.StaticFilters;
+import mage.game.Game;
+import mage.game.permanent.Permanent;
+import mage.game.permanent.token.SpiritTeferiToken;
+import mage.players.Player;
+import mage.target.common.TargetControlledPermanent;
+import mage.target.common.TargetOpponent;
+
+/**
+ *
+ * @author weirddan455
+ */
+public final class TeferiTemporalPilgrim extends CardImpl {
+
+    public TeferiTemporalPilgrim(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.PLANESWALKER}, "{3}{U}{U}");
+
+        this.addSuperType(SuperType.LEGENDARY);
+        this.subtype.add(SubType.TEFERI);
+        this.setStartingLoyalty(4);
+
+        // Whenever you draw a card, put a loyalty counter on Teferi, Temporal Pilgrim.
+        this.addAbility(new DrawCardControllerTriggeredAbility(new AddCountersSourceEffect(CounterType.LOYALTY.createInstance()), false));
+
+        // 0: Draw a card.
+        this.addAbility(new LoyaltyAbility(new DrawCardSourceControllerEffect(1), 0));
+
+        // -2: Create a 2/2 blue Spirit creature token with vigilance and "Whenever you draw a card, put a +1/+1 counter on this creature."
+        this.addAbility(new LoyaltyAbility(new CreateTokenEffect(new SpiritTeferiToken()), -2));
+
+        // -12: Target opponent chooses a permanent they control and returns it to its owner's had. Then they shuffle each nonland permanent they control into its owner's library.
+        Ability ability = new LoyaltyAbility(new TeferiTemporalPilgrimEffect(), -12);
+        ability.addTarget(new TargetOpponent());
+        this.addAbility(ability);
+    }
+
+    private TeferiTemporalPilgrim(final TeferiTemporalPilgrim card) {
+        super(card);
+    }
+
+    @Override
+    public TeferiTemporalPilgrim copy() {
+        return new TeferiTemporalPilgrim(this);
+    }
+}
+
+class TeferiTemporalPilgrimEffect extends OneShotEffect {
+
+    public TeferiTemporalPilgrimEffect() {
+        super(Outcome.Removal);
+        this.staticText = "Target opponent chooses a permanent they control and returns it to its owner's had. Then they shuffle each nonland permanent they control into its owner's library.";
+    }
+
+    private TeferiTemporalPilgrimEffect(final TeferiTemporalPilgrimEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public TeferiTemporalPilgrimEffect copy() {
+        return new TeferiTemporalPilgrimEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player opponent = game.getPlayer(source.getFirstTarget());
+        if (opponent == null) {
+            return false;
+        }
+        TargetControlledPermanent target = new TargetControlledPermanent();
+        target.setNotTarget(true);
+        opponent.chooseTarget(Outcome.ReturnToHand, target, source, game);
+        Permanent toHand = game.getPermanent(target.getFirstTarget());
+        if (toHand != null) {
+            opponent.moveCards(toHand, Zone.HAND, source, game);
```

Comment:
I believe you should be applying the first part of the effect before moving on since it's worded as "Do x. THEN do y". The other cards that get shuffled into the library should see the first one be returned to hand.

Reply:
I'm not sure what you mean by that.

Reply:
Sorry, let me try again. The -12 effect has 2 parts. The first moves a permanent from the battlefield to the hand. The second effect more the other permanents to their owner's libraries.

Because of how the effect is worded they should be happening sequentially so that the permanent returned to the hand is already in the hand and can't see any of the other permanents leave the battlefield.

Though I am less sure if that's an issue here if you are calling the moveCard functions directly.

Reply:
Yes, I understand that.  I don't know what you're asking me to change in the code.

Reply:
Had to look into it a little (see https://github.com/magefree/mage/pull/9304#discussion_r927718153)

Call `game.applyEffects()`

Reply:
That should be game.getState().processAction().

Reply:
Added `game.getState().processAction()` in latest commit.  Also changed the owner HashSet to store UUID instead of Player.  Probably doesn't matter too much but I think it should be easier to hash and less likely to have bugs (Player concedes mid resolution or something.)

Reply:
> Added `game.getState().processAction()` in latest commit. Also changed the owner HashSet to store UUID instead of Player. Probably doesn't matter too much but I think it should be easier to hash and less likely to have bugs (Player concedes mid resolution or something.)

Yes that's also a good idea! I missed that issue.